parlai/core/torch_generator_agent.py

```diff
@@ -1420,20 +1488,26 @@ def advance(self, logprobs):
                 self.context_block_ngram, logprobs, self.context
             )
 
-        hyp_ids, tok_ids, self.scores = self.select_paths(
-            logprobs, self.scores, current_length
-        )
+        # hyp_ids, tok_ids, self.scores, next_choice_token_scores, next_choice_token_ranks
+        path_selection = self.select_paths(logprobs, self.scores, current_length)
+        self.scores = path_selection.scores
         # use clone() here to ensure that self.all_scores will not be changed
         # later due to any penalties to self.scores
         self.all_scores.append(self.scores.clone())
 
-        self.outputs.append(tok_ids)
-        self.bookkeep.append(hyp_ids)
-        tok_id_list = tok_ids.tolist()
+        self.outputs.append(path_selection.token_ids)
+        self.bookkeep.append(path_selection.hypothesis_ids)
+        tok_id_list = path_selection.token_ids.tolist()
         self.partial_hyps = [
-            self.partial_hyps[hyp_ids[i]] + [tok_id_list[i]]
+            self.partial_hyps[path_selection.hypothesis_ids[i]] + [tok_id_list[i]]
             for i in range(self.beam_size)
         ]
+        self.token_scores = torch.cat(
```

Comment:
Won't it be a little faster this way because it keeps the score on the GPU?

Reply:
I don't necessarily see why it would be faster because we'll have to move it off GPU eventually and AFAICT we are not doing any operations which would be faster on GPU, nevertheless I don't have a strong opinion on this so I will defer to what you think is best here for future work